aws/aws.go

```diff
@@ -393,6 +395,52 @@ func GetAllResources(targetRegions []string, excludeAfter time.Time, resourceTyp
 		}
 		// End RDS DB Clusters
 
+		// S3 Buckets
+		s3Buckets := S3Buckets{}
+		if IsNukeable(s3Buckets.ResourceName(), resourceTypes) {
+			var bucketNamesPerRegion map[string][]*string
+
+			// AWS S3 buckets list operation lists all buckets irrespective of regions.
+			// For each bucket we have to make a separate call to find the bucket region.
+			// Hence for x buckets and a total of y target regions - we need to make:
+			// (x + 1) * y calls i.e. 1 call to list all x buckets, x calls to find out
+			// each bucket's region and repeat the process for each of the y regions.
+
+			// getAllS3Buckets returns a map of regions to buckets and we call it only once -
+			// thereby reducing total calls from (x + 1) * y to only (x + 1) for the first region -
+			// followed by a cache lookup for rest of the regions.
+			if _, ok := resourcesCache["S3"]; !ok {
+				bucketNamesPerRegion, err = getAllS3Buckets(session, excludeAfter)
+				if err != nil {
+					return nil, errors.WithStackTrace(err)
+				}
+
+				if resourcesCache["S3"] == nil {
+					resourcesCache["S3"] = make(map[string][]*string)
+				}
+
+				for k := range bucketNamesPerRegion {
+					resourcesCache["S3"][k] = bucketNamesPerRegion[k]
+				}
+			} else {
+				// Cache lookup for subsequent calls
+				bucketNamesPerRegion = resourcesCache["S3"]
+			}
+
+			// getAllS3Buckets returns only those regions which have buckets
+			if _, ok := bucketNamesPerRegion[region]; !ok {
+				bucketNamesPerRegion[region] = []*string{}
+			}
+
+			bucketNames := bucketNamesPerRegion[region]
+
+			if len(bucketNamesPerRegion[region]) > 0 {
+				s3Buckets.Names = awsgo.StringValueSlice(bucketNames)
+				resourcesInRegion.Resources = append(resourcesInRegion.Resources, s3Buckets)
+			}
```

Comment:
This was added because ```bucketNamesPerRegion``` contains only those regions which have buckets. When we are iterating through all AWS regions in the outer loop - we might encounter a region which has no bucket hence no entry in ```bucketNamesPerRegion``` - in that case ```bucketNamesPerRegion[region]``` will fail. But I will check it again and verify if that is the case.